internal/deviations/deviations.go

```diff
@@ -65,27 +65,91 @@ package deviations
 
 import (
 	"flag"
+	"fmt"
+	"regexp"
 
 	log "github.com/golang/glog"
 	"github.com/openconfig/featureprofiles/internal/metadata"
 	mpb "github.com/openconfig/featureprofiles/proto/metadata_go_proto"
 	"github.com/openconfig/ondatra"
 )
 
-// lookupDutDeviations returns the deviations for the specified dut.
-func lookupDUTDeviations(dut *ondatra.DUTDevice) *mpb.Metadata_Deviations {
+func matchRegex(regex string, deviceInfo string) bool {
+	match, err := regexp.MatchString(regex, deviceInfo)
+	if err != nil {
+		log.Errorf("Error with regex match %v", err)
+	}
+	return match
+}
+
+func isPlatformExceptionsValid(platformExceptions *mpb.Metadata_PlatformExceptions) error {
+	if platformExceptions.GetPlatform().Vendor.String() == "" {
+		return fmt.Errorf("vendor should be specified in textproto %v", platformExceptions)
+	}
+	if platformExceptions.GetPlatform().GetHardwareModelRegex() != "" && len(platformExceptions.GetPlatform().GetHardwareModel()) > 0 {
+		return fmt.Errorf("vendor should be specified in textproto %v", platformExceptions)
+	}
+	return nil
+}
+
+func lookupDeviations(dut *ondatra.DUTDevice) (*mpb.Metadata_PlatformExceptions, error) {
+	matchedPlatformException := &mpb.Metadata_PlatformExceptions{}
+	isPlatformExceptionsMatched := false
+
 	for _, platformExceptions := range metadata.Get().PlatformExceptions {
+		if err := isPlatformExceptionsValid(platformExceptions); err != nil {
+			return nil, err
+		}
 		if dut.Device.Vendor().String() != platformExceptions.GetPlatform().Vendor.String() {
 			continue
 		}
+		// Exit test if both hardware_model_regex and hardware_model fields are set.
+		if hardwareModelRegex := platformExceptions.GetPlatform().GetHardwareModelRegex(); hardwareModelRegex != "" {
+			// If hardware_model_regex is set and does not match, continue
+			if !matchRegex(hardwareModelRegex, dut.Device.Model()) {
+				continue
+			}
+		}
+		if softwareVersionRegex := platformExceptions.GetPlatform().GetSoftwareVersionRegex(); softwareVersionRegex != "" {
+			// If software_version_regex is set and does not match, continue
+			if !matchRegex(softwareVersionRegex, dut.Device.Version()) {
+				continue
+			}
+		}
+
+		// TODO(prinikasn): Refactor after hardware_model field is removed.
+		if len(platformExceptions.GetPlatform().GetHardwareModel()) == 0 {
+			if isPlatformExceptionsMatched {
+				return nil, fmt.Errorf("cannot have more than one match within platform_exceptions fields %v and %v", matchedPlatformException, platformExceptions)
+			}
+			matchedPlatformException = platformExceptions
```

Comment:
It would be more consistent with the prior checks if you continue when there is a non-match, and then you only set this at the end of the loop (if there were no prior continues)

Reply:
done